### PR TITLE
Contour lines

### PR DESCRIPTION
**Pull request by dgirardeau**
- make the "contour lines" computation algorithm separate from the Rasterize tool
- update the Cross Section tool to be able to compute contours of a point cloud from a slice or from multiple slices

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -84,6 +84,8 @@ v2.11 (Anoia) - (in development)
 		- option to select the extracted contour type (LOWER, UPPER or FULL)
 		- 'up' direction is always Z for slices normal to X or Y (the local X, Y or Z directions of the active cross-section)
 		- 'up' direction is X for slices normal to Z
+		- ability to extract the real contours of the points inside each slice (single slice mode or 'repeat' mode)
```
Comment (asmaloney) [inline CHANGELOG.md:87]: These changes should go in a new 2.12 section.

File: qCC/ccClippingBoxRepeatDlg.cpp [hunk 2]
```diff
@@ -23,31 +23,32 @@
...
 		connect(yRepeatCheckBox, &QAbstractButton::toggled, this, &ccClippingBoxRepeatDlg::onDimYChecked);
 		connect(zRepeatCheckBox, &QAbstractButton::toggled, this, &ccClippingBoxRepeatDlg::onDimZChecked);
+		
 		setFlatDim(0);
 
-		extractContoursGroupBox->setChecked(true);
-		extractContoursGroupBox->setCheckable(false);
-		projectOnBestFitCheckBox->setVisible(true);
-		projectOnBestFitCheckBox->setChecked(false);
+		//extractEnvelopesGroupBox->setChecked(true); //DGM: we now have the choice between envelope and contours!
```
Comment (asmaloney) [inline qCC/ccClippingBoxRepeatDlg.cpp:48]: Remove commented-out code

File: qCC/ccClippingBoxRepeatDlg.cpp [hunk 3]
```diff
@@ -99,17 +116,20 @@ void ccClippingBoxRepeatDlg::onDimChecked(bool)
 
 	if (sum == 1)
 	{
-		if (!projectOnBestFitCheckBox->isVisible())
-			projectOnBestFitCheckBox->setChecked(false);
-		projectOnBestFitCheckBox->setVisible(true);
-		contourTypeComboBox->setEnabled(true);
+		if (!envProjectPointsOnBestFitCheckBox->isVisible())
+			envProjectPointsOnBestFitCheckBox->setChecked(false);
```
Comment (asmaloney) [inline qCC/ccClippingBoxRepeatDlg.cpp:120]: Missing '{' and '}'

File: qCC/ccClippingBoxTool.cpp [hunk 4]
```diff
@@ -29,17 +30,18 @@
 
 //qCC_db
 #include <ccClipBox.h>
+#include <ccRasterGrid.h>
```
Comment (asmaloney) [inline qCC/ccClippingBoxTool.cpp:33]: Should be alphabetical.

File: qCC/ccClippingBoxTool.cpp [hunk 5]
```diff
@@ -530,21 +532,31 @@ static unsigned ComputeGridDimensions(	const ccBBox& localBox,
 	return cellCount;
 }
 
+static constexpr char* s_originEntityUUID = "OriginEntityUUID";
```
Comment (asmaloney) [inline qCC/ccClippingBoxTool.cpp:535]: For modern C++, statics are replaced by using in an anonymous namespace (preferably at the top of the file). Things in anonymous namespaces are only valid within its translation unit.

File: qCC/ccClippingBoxTool.cpp [hunk 6]
```diff
@@ -909,13 +953,174 @@ bool ccClippingBoxTool::ExtractSlicesAndContours
...
-		//extract contour polylines (optionaly)
-		if (!error && extractContours && cloudSliceCount != 0)
+		//extract level set (optionaly)
+		if (!error && extractLevelSet && cloudSliceCount != 0)
+		{
+			for (int iteration = 0; iteration < 1; ++iteration) //fake loop for easy break
+			{
+				if (progressDialog)
+				{
+					progressDialog->setWindowTitle("Level set extraction");
```
Comment (asmaloney) [inline qCC/ccClippingBoxTool.cpp:963]: tr() ?

File: qCC/ccClippingBoxTool.cpp [hunk 7]
```diff
@@ -909,13 +953,174 @@ bool ccClippingBoxTool::ExtractSlicesAndContours
...
+					{
+						ccLog::Warning(QString("Failed to generate contour lines for cloud #%1").arg(i + 1));
+					}
+
+					if (progressDialog)
+					{
+						if (progressDialog->wasCanceled())
+						{
+							error = true;
+							ccLog::Warning(QString("[ExtractSlicesAndContours] Process canceled by user"));
```
Comment (asmaloney) [inline qCC/ccClippingBoxTool.cpp:1108]: tr() ?

File: qCC/ccClippingBoxTool.cpp [hunk 8]
```diff
@@ -909,13 +953,174 @@ bool ccClippingBoxTool::ExtractSlicesAndContours
...
+		}
+
+		//extract envelopes as polylines (optionaly)
+		if (!error && extractEnvelopes && cloudSliceCount != 0)
 		{
 			if (progressDialog)
 			{
-				progressDialog->setWindowTitle("Contour extraction");
-				progressDialog->setInfo(QObject::tr("Contour(s): %L1").arg(cloudSliceCount));
+				progressDialog->setWindowTitle("Envelope extraction");
```
Comment (asmaloney) [inline qCC/ccClippingBoxTool.cpp:1122]: tr() ?

File: qCC/ccClippingBoxTool.cpp [hunk 9]
```diff
@@ -1196,67 +1465,292 @@ void ccClippingBoxTool::extractSlicesAndContours(bool extractSlices, bool extrac
...
 	}
 
 	if (m_associatedWin)
+	{
 		m_associatedWin->redraw();
+	}
+
+	if (sliceCount != 0)
+	{
+		QMessageBox::warning(nullptr, "Process finished", QString("%1 slices have been generated.\n(you may have to close the tool and hide the initial cloud to see them...)").arg(sliceCount));
```
Comment (asmaloney) [inline qCC/ccClippingBoxTool.cpp:1748]: tr() ?

File: qCC/ccClippingBoxTool.cpp [hunk 10]
```diff
@@ -1196,67 +1465,292 @@ void ccClippingBoxTool::extractSlicesAndContours(bool extractSlices, bool extrac
...
+	}
+	else if (s_extractSliceCloudsOrMeshes)
+	{
+		QMessageBox::warning(nullptr, "Process finished", "The process has generated no output");
```
Comment (asmaloney) [inline qCC/ccClippingBoxTool.cpp:1752]: tr() ?

File: qCC/ccContourLinesGenerator.h [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+//#  This program is free software; you can redistribute it and/or modify  #
+//#  it under the terms of the GNU General Public License as published by  #
+//#  the Free Software Foundation; version 2 or later of the License.      #
+//#                                                                        #
+//#  This program is distributed in the hope that it will be useful,       #
+//#  but WITHOUT ANY WARRANTY; without even the implied warranty of        #
+//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the          #
+//#  GNU General Public License for more details.                          #
+//#                                                                        #
+//#          COPYRIGHT: EDF R&D / TELECOM ParisTech (ENST-TSI)             #
```
Comment (asmaloney) [inline qCC/ccContourLinesGenerator.h:16]: "CloudCompare Project"?

Comment (dgirardeau) [reply qCC/ccContourLinesGenerator.h:16]: Well this is just a copy-paste from existing code (in ccRasterizeTool), so I preferred to keep the original copyright notice.

**Pull request merged**